### PR TITLE
add stable IDs to messages

### PR DESCRIPTION
**Pull request by mcginty**
Adds back all those fun animations and should help solve scrolling issues since the RecyclerView can be much more intelligent about data changes now.

### CODE REVIEW

**Comment by mcginty** [comment]
updated

File: src/org/thoughtcrime/securesms/ConversationAdapter.java [hunk 1]
```diff
@@ -171,6 +199,14 @@ public int getItemViewType(@NonNull Cursor cursor) {
     }
   }
 
+  @Override
+  public long getItemId(@NonNull Cursor cursor) {
+    digest.reset();
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationAdapter.java]: this isn't necessary if you never call update

File: src/org/thoughtcrime/securesms/ConversationAdapter.java [hunk 2]
```diff
@@ -201,4 +237,14 @@ public void clearSelection() {
...
   }
+
+  private long firstBytesAsLong(byte[] bytes) {
+    long l = 0;
+    for(int i = 0; i < Long.SIZE / 8; i++) {
+      l <<= 8;
+      l ^= bytes[i] & 0xFF;
+    }
+    return l;
+  }
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationAdapter.java]: Drop this and replace with Conversions.byteArrayToLong

File: src/org/thoughtcrime/securesms/ConversationAdapter.java [hunk 3]
```diff
@@ -106,6 +128,12 @@ public ConversationAdapter(@NonNull Context context,
     this.recipients    = recipients;
     this.inflater      = LayoutInflater.from(context);
     this.db            = DatabaseFactory.getMmsSmsDatabase(context);
+    try {
+      digest = MessageDigest.getInstance("SHA-1");
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationAdapter.java]: "SHA1"

File: src/org/thoughtcrime/securesms/ConversationAdapter.java [hunk 4]
```diff
@@ -106,6 +128,12 @@ public ConversationAdapter(@NonNull Context context,
...
+    } catch (NoSuchAlgorithmException nsae) {
+      throw new IllegalStateException("SHA-1 isn't supported!");
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationAdapter.java]: assertion error

**Pull request closed without merging**